clang/lib/AST/Type.cpp

```diff
@@ -5016,3 +5024,254 @@ void AutoType::Profile(llvm::FoldingSetNodeID &ID, const ASTContext &Context) {
+
+  default:
+    break;
+  }
+  return false;
+}
+
+bool FunctionEffect::diagnoseFunctionCall(bool Direct,
+                                          FunctionEffectSet CalleeFX) const {
+  switch (type()) {
+  case Type::NonAllocating:
+  case Type::NonBlocking: {
+    const Type CallerType = type();
+    for (const auto &Effect : CalleeFX) {
+      const Type ET = Effect.type();
+      // Does callee have same or stronger constraint?
+      if (ET == CallerType ||
+          (CallerType == Type::NonAllocating && ET == Type::NonBlocking)) {
+        return false; // no diagnostic
+      }
+    }
+    return true; // warning
+  }
+  default:
+    break;
+  }
+  return false;
+}
+
+// =====
+
+MutableFunctionEffectSet::MutableFunctionEffectSet(
+    const FunctionEffect &effect) {
+  push_back(effect);
+}
+
+void MutableFunctionEffectSet::insert(const FunctionEffect &Effect) {
+  const auto &Iter = std::lower_bound(begin(), end(), Effect);
+  if (Iter == end() || *Iter != Effect) {
+    insert(Iter, Effect);
+  }
+}
+
+MutableFunctionEffectSet &
+MutableFunctionEffectSet::operator|=(FunctionEffectSet RHS) {
+  // TODO: For large RHS sets, use set_union or a custom insert-in-place
+  for (const auto &Effect : RHS) {
+    insert(Effect);
+  }
+  return *this;
+}
+
+FunctionEffectSet FunctionEffectSet::getUnion(ASTContext &C,
+                                              const FunctionEffectSet &LHS,
+                                              const FunctionEffectSet &RHS) {
+  // Optimize for one of the two sets being empty
+  if (LHS.empty())
+    return RHS;
+  if (RHS.empty())
+    return LHS;
+
+  // Optimize the case where the two sets are identical
+  if (LHS == RHS)
+    return LHS;
+
+  MutableFunctionEffectSet Vec;
+  Vec.reserve(LHS.size() + RHS.size());
+  std::set_union(LHS.begin(), LHS.end(), RHS.begin(), RHS.end(),
+                 std::back_inserter(Vec));
+  // The result of a set operation is an ordered/unique set.
+  return C.getUniquedFunctionEffectSet(Vec);
+}
+
+MutableFunctionEffectSet
+FunctionEffectSet::operator&(const FunctionEffectSet &RHS) const {
+  const FunctionEffectSet &LHS = *this;
+  if (LHS.empty() || RHS.empty()) {
+    return {};
+  }
+
+  MutableFunctionEffectSet Vec;
+  std::set_intersection(LHS.begin(), LHS.end(), RHS.begin(), RHS.end(),
+                        std::back_inserter(Vec));
+  // The result of a set operation is an ordered/unique set.
+  return Vec;
+}
+
+// TODO: inline?
+FunctionEffectSet FunctionEffectSet::get(QualType QT) {
+  if (QT->isReferenceType())
+    QT = QT.getNonReferenceType();
+  if (QT->isPointerType())
+    QT = QT->getPointeeType();
+
+  if (const auto *BT = QT->getAs<BlockPointerType>()) {
+    QT = BT->getPointeeType();
+  } else if (const auto *MP = QT->getAs<MemberPointerType>()) {
+    if (MP->isMemberFunctionPointer()) {
+      QT = MP->getPointeeType();
+    }
+  }
+
+  if (const auto *FPT = QT->getAs<FunctionProtoType>())
+    return FPT->getFunctionEffects();
```

Comment:
*All of this* seems a bit brittle to me; I think just requiring the caller to instead dig out the `FunctionProtoType` themselves and then call `getFunctionEffects()` would probably be a better idea. Because this function seems to be for the case of ‘I have this random thing that I know contains a function type *somewhere*, but I have no idea where or what it is beyond that’, which seems a bit unlikely—is there any case where we actually need this and can’t just access the `FunctionProtoType` directly?

Reply:
The common theme here is having to peel off references, pointers, etc. in order to get to the `FunctionProtoType`. I too am unhappy that it seems brittle. But at least it's not repeated into its callers, which are:

`Sema::diagnoseFunctionEffectConversion` which only has a pair of `QualType`s. This is only called from `Sema::ImpCastExprToType`, which only has a `QualType` and an `Expr*`.

In the analysis phase, we can have a `ValueDecl` where adoption in a large codebase revealed the need for all of these situations. The difficulties arise with patterns where a function is being called indirectly, and the function/block/PTMF might itself be part of a cast expression, or a variable that is a reference or pointer.

I would hope this problem has been solved elsewhere and I just didn't know where to look.

Reply:
This is now:
```
FunctionEffectsRef FunctionEffectsRef::get(QualType QT) {
  if (QualType Pointee = QT->getPointeeType(); !Pointee.isNull())
    QT = Pointee;
  if (const auto *FPT = QT->getAs<FunctionProtoType>())
    return FPT->getFunctionEffects();
  return {};
}
```